### PR TITLE
WIP: Add initial code for backend

### PR DESCRIPTION
**Pull request by slakh96**
Added initial code for the backend - mongo db connection, db functions, a server and a program to populate the mongo db

### CODE REVIEW

**Comment by slakh96** [comment]
Note: I have tested that the backend code, including initialization script, server routes etc work using postman, but we probably want to eventually write unittests when we actually get something meaningful in these files.

File: backend/db_interface.py
Comment (sultansidhu) [inline backend/db_interface.py]: please check for KeyError and have some kind of a default value. Can be done in one line by doing `output.get('data', 'default_val')`

File: backend/populate_db.py
Comment (sultansidhu) [inline backend/populate_db.py]: Is this file to be committed? or is this a file for testing?

File: backend/requirements.txt
Comment (sultansidhu) [inline backend/requirements.txt]: what is cryptography required for?

File: backend/requirements.txt
Comment (sultansidhu) [inline backend/requirements.txt]: same with all these?

File: backend/requirements.txt
Comment (sultansidhu) [inline backend/requirements.txt]: and many more... that's a lot of requirements for a small code base damn 😳

File: backend/db_connection.py
Comment (sultansidhu) [inline backend/db_connection.py]: Please error check with a default value, use a one-liner like `connection.get('verwiki', default_val)`

File: README.md
Comment (sultansidhu) [inline README.md]: Please mention creating a Conda environment and development in that. 
The following steps should be helpful:

`conda create --name <env name>`
`conda activate <env name>`
And then `pip install -r requirements.txt`

File: backend/server.py
Comment (sultansidhu) [inline backend/server.py]: check for KeyError here as well 😄

**Review by sultansidhu — CHANGES_REQUESTED** [review]

File: README.md
Comment (sultansidhu) [reply README.md]: Made the README changes in my own CR btw, so don't need to do this anymore.

File: backend/populate_db.py
Comment (slakh96) [reply backend/populate_db.py]: committed; it is for you guys to populate your local dbs

File: backend/requirements.txt
Comment (slakh96) [reply backend/requirements.txt]: yeah these are probably nested dependencies; by doing `pip freeze` it initializes the requirements.txt, with all the current requirements including nested dependencies...can you try on your end to generate requirements and see what it comes up with?

File: backend/db_connection.py
Comment (slakh96) [reply backend/db_connection.py]: no error check would be required here, since it will create the DB if it does not already exist (e.g. when you guys initially get set up)

File: backend/server.py
Comment (slakh96) [reply backend/server.py]: If the client didn't give a body with ID, this automatically throws a 500 error...do we want something more specific?

File: backend/db_connection.py
Comment (slakh96) [reply backend/db_connection.py]: Actually, let's future proof this as it will be needed later; updated the code

File: backend/server.py
Comment (slakh96) [reply backend/server.py]: Made this a lot more specific, please check the new error handling

File: backend/db_interface.py
Comment (slakh96) [reply backend/db_interface.py]: Added error checking for this section; we should never run into this particular error as things should be validated before entering db, but it doesn't hurt to check

**Comment by slakh96** [comment]
Don't merge this yet: I want to add linting and a couple of unittests to this MR

File: backend/requirements.txt
Comment (sultansidhu) [reply backend/requirements.txt]: I tried pip install -r and it runs into an error. I think it would be best to initialize a Conda env, and manually putting all the things in the requirements. using pip freeze generates lots of nested dependencies, which may not be available in the PyPI repositories.

File: backend/db_interface.py
Comment (sultansidhu) [inline backend/db_interface.py:4]: lets be consistent and use either single quote or double quote strings throughout.

File: backend/db_interface.py
Comment (sultansidhu) [inline backend/db_interface.py]: why raising KeyError here? 
Also, when trying to put dynamic values in strings, lets use f-strings (they make code very elegant, and get rid of excessively long lines by taking out + signs and str() calls)
The f-string would be 
`f"No tree found with the id {id}"`

File: backend/db_interface.py
Comment (sultansidhu) [inline backend/db_interface.py]: This here should be KeyError no? Because a Key is not present in a Key-Value relation. Can also get rid of print if raising Exception with a print val imo?

File: backend/db_interface.py
Comment (sultansidhu) [inline backend/db_interface.py]: Don't need print statement as already printing prompt when throwing error imo

File: backend/db_interface.py
Comment (sultansidhu) [reply backend/db_interface.py]: Also, f-strings please : )

**Review by sultansidhu — CHANGES_REQUESTED** [review]

File: backend/db_interface.py
Comment (slakh96) [reply backend/db_interface.py:4]: Can we make this part of the linting step of the pipeline? Otherwise it's a losing battle

File: backend/db_interface.py
Comment (slakh96) [reply backend/db_interface.py]: Here there is a key error since there was no tree found matching the given ID, i.e. the key was not in the DB. See here: https://realpython.com/python-keyerror/

File: backend/db_interface.py
Comment (slakh96) [reply backend/db_interface.py]: At this point we were already able to find an object matching the key, so it shouldn't be a KeyError. In this case, the object in the DB doesn't follow the expected format, hence the TypeError. We also should have a print here, since it is an immediate cause for concern if our DB has malformed data (Data should be verified before entering DB)

yeah, f-strings are equivalent to what I have here, but I guess I can change it; if we feel strongly about f-strings they should be in the linter

File: backend/db_interface.py
Comment (slakh96) [reply backend/db_interface.py]: Right, this is just code for testing purposes

File: backend/db_interface.py
Comment (slakh96) [reply backend/db_interface.py:4]: Nevermind, turns out that this step is already part of the `black` linter

File: backend/requirements.txt
Comment (slakh96) [reply backend/requirements.txt]: Ok, looks like you already fixed requirements.txt

File: backend/server.py
Comment (mit1221) [inline backend/server.py]: The tree id should be part of the route url, instead of the request body. The route should probably be something like: "/trees/<int:id>"

File: backend/server.py
Comment (mit1221) [inline backend/server.py:2]: Since this will be a REST server and not a web server, we should use flask_restful rather than vanilla flask.

**Review by sultansidhu — APPROVED** [review]
Looks good now! Linter on backend is failing, after fixing that can merge safely : ) great work!!!

File: backend/db_interface.py
Comment (sultansidhu) [reply backend/db_interface.py]: They are equivalent, but form a much cleaner code style which abstracts out the str() call and the plus signs.

File: backend/server.py
Comment (slakh96) [reply backend/server.py:2]: should discuss this in a future meeting when we move on to phase 1 - professor

File: backend/db_interface.py
Comment (slakh96) [reply backend/db_interface.py]: ok, though probably not worth bothering about; do you know of any linter step that does this?

File: backend/server.py
Comment (slakh96) [reply backend/server.py]: This makes the code a lot shorter, ty!

**Review by mit1221 — APPROVED** [review]

**Pull request merged**
